rfcs/0119-testing-conventions.md

```diff
@@ -0,0 +1,105 @@
+an unambiguous way to say that a given package, and all downstream dependencies, have
+had as many automated test ran possible. This will give a high degree of certainty that
+a given change is less likely to manifest regressions once introduced on a release
+channel.
+
+Another desire of this rfc is also to have a way for various review tools
+(e.g. ofborg, hydra, nixpkgs-review) to have a standard way to determine if a
+package has additional tests which can help verify its correctness.
+
+# Motivation
+[motivation]: #motivation
+
+Breakages are a constant painpoint for nixpkgs. It is a very poor user experience to
+have a configuration broken because one or more packages fail to build. Often when
+these breakages occur, it is because the change had a large impact on the entirety
+of nixpkgs; and unless there's a dedicated hydra jobset for the pull request, it's
+infeasible to expect pull request authors to verify every package affected
+by a change they are proposing. However, it is feasible to specify packages that
+are very likely to be affected by changes in another package, and use this information
+to help mitigate regressions from appearing in release channels.
+
+# Detailed design
+[design]: #detailed-design
+
+Standardize `passthru.tests.<name>` and `passthru.nixosTests.<name>` as a mechanism of 
+more expensive but automatic testing for nixpkgs. As well as encourage the usage of
+`checkPhase` or `installCheckPhase` when packaging within nixpkgs.
+
+Usage for `passthru.nixosTests.<name>`
+- Reserved for tests utilitizing the nixosTest utilties.
+  - Generally these are more resource intensive, and may require additional system features
+  such as kvm
+
+Usage for `passthru.tests.<name>`:
+- Running tests which include downstream dependencies.
+  - This avoids cyclic dependency issues for test suites.
+- Running lengthy or more resource expensive tests.
+  - There should be a priority on making package builds as short as possible.
+  - This reduces the amount of compute required for everyone reviewing, building, or iterating on packages.
+- Referencing downstream dependencies which are most likely to experience regressions.
+  - Most applicable to [RFC 0088 - Nixpkgs Breaking Change Policy](https://github.com/NixOS/rfcs/pull/88),
+as this will help define what breakages a pull request author should take ownership.
+
+Usage for mkDerivation's `checkPhase`:
+- Quick "cheap" tests, which run units tests and maybe some addtional scenarios.
+- Since this contributes to the "build time" of a package, there should be some
+emphasis on ensuring this phase isn't bloated.
+
+Usage for mkDerivations `installCheckPhase`:
+- A quick trivial example (e.g. `<command> --help`) to demonstrate that one (or more)
+of the programs were linked correctly.
+- Assert behavior post installation (e.g. python's native extensions only get installed
+and are not present in a build directory)
+
```

Comment:
some programs are hard to test automatically so how about creating a new meta attribute like `testingInstructions`

Reply:
I strongly oppose this idea.

IIUC the goal of this RFC is to make it easier for changes to upstream packages to be tested. The end goal is that we have automatic tooling that can test packages, notify maintainers of breakages and eventually mark as broken if the maintainers are unable to fix the package in time. Adding required manual testing puts unacceptable levels of burden on core package maintainers (that are depended on by hundreds or thousands of other packages).

I think a `testingInstructions` attribute may be an interesting and useful idea but I think it would serve a different purpose as the formalized testing specified by this RFC.

TL;DR I don't want to require people to manually test loads of packages, if you want your package not to break due to changes in dependencies you need automated tests.

Reply:
I guess `testingInstructions` would fit here if it came with «replacing this with a test that looks robust will be accepted even if the test implementation is ugly», but I do not believe such a commitment would be accepted into the attribute semantics…

Reply:
Perhaps we could record merge conditions instead?
Some packages don't need manual testing, so all that's needed for a merge is a review of the changelog, but this information has not been recorded yet.

Reply:
I would like to have a `CONTRIBUTING.md` in a packages directory with documentation for package maintainers and contributors, like how to test, how to update...
In addition to that, a `README.md` with documentation for users.